pkg/proxy/winkernel/hcn_utils_mock.go

```diff
@@ -0,0 +1,264 @@
+//go:build windows
+// +build windows
+
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package winkernel
+
+import (
+	"errors"
+	"strconv"
+
+	"github.com/Microsoft/hcsshim/hcn"
+)
+
+const (
+	guid                        = "123ABC"
+	destinationPrefix           = "192.168.2.0/24"
+	isolationId                 = 4096
+	providerAddress             = "10.0.0.3"
+	distributedRouterMacAddress = "00-11-22-33-44-55"
+	macRangeStartAddress        = "00-15-5D-52-C0-00"
+	macRangeEndAddress          = "00-15-5D-52-CF-FF"
+	ipAddressPrefix             = "192.168.1.0/24"
+	subnetNextHop               = "192.168.1.1"
+	subnetDestinationPrefix     = "0.0.0.0/0"
+	lbIDPrefix                  = "MOCK-LB-ID-"
+	epIDPrefix                  = "MOCK-EP-ID-"
+)
+
+// mockHCN saves the created endpoints and loadbalancers in slices to be able to work with them easier
+type mockHCN struct {
```

Comment:
Is there a reason to mock all the hcnUtils? To my understanding all of these are helpers that ultimately call into HCN.

Would it perhaps be better to only mock all of the hcsshim/HCN signature that the hcnUtils use? Store any of the created endpoints/loadbalancers as part of the "HCN" mock in memory and remove the structs/interfaces that mock hcnUtils. Tests and proxier code would use the same hcnUtils code. 
 
hcnUtils for tests would initialize the hcn mock package which store artificial endpoints/loadbalancers, and real proxier code would use hcnUtils that initializes/references real hcn package.

Maybe I am missing something but feels to me this might reduce a layer of abstraction and make the code easier to understand, without reducing robustness of tests. Curious to hear your thoughts?

Reply:
hcnUtils is the final place where the test cases have control to mock the function call in the existing implementation. Considering an example of network.createEndpoint():

**network.createEndpoint()** -Calls-> **createEndpoint** [Which is in hcn package of vendor directory] -Calls-> **hcnOpenNetwork** [Which is in hcn package of vendor directory] -Calls-> **syscall.Syscall**

If you have a look at the above call flow, test code dont have any handle to avoid the syscall. That's the reason behind mocking hncUtils.